### PR TITLE
Add flatpak build manifest and necessary extra data

### PR DESCRIPTION
**Pull request by fpiesche**
There's some TODOs here - feedback on these would be good! Other than that, this should be fairly directly in a shape where it's submissible to Flathub :)

### CODE REVIEW

File: net.shadps4.shadps4.yml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        sha256: 7fb9f29b86f6842a6f67ebdf6b56b3fc25671b3bca50448e1505f47f37eb19d5
+        x-checker-data:
+          type: json
+          url: https://api.github.com/repos/tencent/rapidjson/releases/latest
+          version-query: .tag_name
+          url-query: .zipball_url
+    cleanup:
+      - '*'
+
+  # TODO: libfmt is included as a submodule in externals - do we need to build it separately?
```
Comment (ngoquang2708) [inline net.shadps4.shadps4.yml:66]: The KDE Runtime 6.8 is currently missing ```libfmt```. There is a merged PR to add ```libfmt``` to the FreeDesktop Runtime, which KDE Runtime based on. I don't know how long until the KDE Runtime get updated.
At the mean time, we can use something similar to this
```
    post-install:
      # FIXME Temporary workaround until libfmt is added to the runtime
      - cp /lib/*-linux-gnu/libfmt* /app/lib/
```

and then we can remove ```libfmt``` from this manifest.

File: net.shadps4.shadps4.yml [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  #           url: https://api.github.com/repos/fmtlib/fmt/releases/latest
+  #           version-query: .tag_name
+  #           url-query: .zipball_url
+  #   cleanup:
+  #     - '*'
+
+  # TODO: Original manifest by deadmeu included libevdev. Do we need this?
+  # TODO: shadps4 seems to build fine without it, even though it's listed as a requirement
+  # TODO: in https://github.com/shadps4-emu/shadPS4/blob/main/documents/building-linux.md
+  - name: evdev
```
Comment (ngoquang2708) [inline net.shadps4.shadps4.yml:87]: I can confirm that this is not needed. AppImages don't include this as well.

**Comment by ngoquang2708** [comment]
Missing this https://github.com/shadps4-emu/shadPS4/pull/1474#issuecomment-2453822651.

File: net.shadps4.shadps4.yml [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  # gamepad and 3D hardware support
+  - --device=dri
+  - --device=input
+  - --allow=bluetooth
+  # Wayland + fallback
+  - --socket=wayland
+  - --socket=fallback-x11
+  # Audio
+  - --socket=pulseaudio
+  # Discord rich presence?
```
Comment (ngoquang2708) [inline net.shadps4.shadps4.yml:22]: https://github.com/shadps4-emu/shadPS4/pull/1474#issuecomment-2454254735

File: net.shadps4.shadps4.yml (hunk 1 above)
Comment (fpiesche) [reply net.shadps4.shadps4.yml:66]: Ah, is it in the submodules just for the headers? I was wondering if it gets built alongside shadps4 if it's not on the system itself since it's included there.

Comment (ngoquang2708) [reply net.shadps4.shadps4.yml:66]: The one in submodules is just a fallback for when the host don't have it installed.

**Comment by fpiesche** [comment]
> How do we handle release/nightly update channels?

Flathub itself is mostly intended for stable releases, and this manifest is geared around that (cf. the x-checker-data segments looking for tagged releases in the shadps4 git repo).

Flathub does have a beta channel that could be used for dev builds; the way this is handled is by adding a `beta` branch to the `flathub/[app id]` repo that's created once the project is submitted; the manifest in that channel will be built and published to the public Flathub beta repository.

File: net.shadps4.shadps4.yml [hunk 4]
```diff
@@ -0,0 +1,118 @@
+app-id: net.shadps4.shadps4
```
Comment (ngoquang2708) [inline net.shadps4.shadps4.yml:1]: I think this should change to this.
```suggestion
app-id: net.shadps4.shadPS4
```
That the way I see other apps used. Applied for other files as well.

**Comment by ngoquang2708** [comment]
> published to the public Flathub beta repository.

So if user want to use the beta channel they have to add the Flathub beta repo manually?

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+.flatpak-builder
```
Comment (ngoquang2708) [inline .gitignore:1]: These would be useful if if you want to build locally for testing purpose.
```suggestion
.flatpak-builder
builddir
repo
*.flatpak
```

**Comment by fpiesche** [comment]
> > published to the public Flathub beta repository.
> 
> So if user want to use the beta channel they have to add the Flathub beta repo manually?

Yup, and install from the beta repo specifically either using the relevant control in their graphical package manager, or via `flatpak install flathub-beta net.shadps4.shadps4`.

File: net.shadps4.shadps4.metainfo.xml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  <developer id="net.shadps4.shadps4">
+    <name>shadPS4 Contributors</name>
+    <url translate="no">https://github.com/shadps4-emu/shadps4/graphs/contributors</url>
+  </developer>
+  <summary>PS4 Emulator</summary>
+  <metadata_license translate="no">CC0-1.0</metadata_license>
+  <project_license translate="no">GPL-2.0</project_license>
+  <launchable type="desktop-id" translate="no">net.shadps4.shadps4.desktop</launchable>
+  <url type="homepage" translate="no">https://shadps4.net/</url>
+  <icon type="remote" width="256" height="256" translate="no">https://cdn.jsdelivr.net/gh/shadps4-emu/shadps4@main/src/images/net.shadps4.shadps4.svg</icon>
```
Comment (ngoquang2708) [inline net.shadps4.shadps4.metainfo.xml:14]: To avoid any frictions that block this PR from getting merged, I think you should update external links to point back to GH or preferably to official website. Then when everything is sorted out. You can make another PR to update for these links.

Comment (fpiesche) [reply net.shadps4.shadps4.metainfo.xml:14]: The jsdelivr links are just CDN mirrors for Github; jsdelivr is a free CDN for open-source projects backed by a number of the major commercial orgs in the CDN space - so any changes to these files in the github repository will be automatically reflected in the jsdelivr CDN cache.

File: net.shadps4.shadps4.yml [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+      - -DENABLE_QT_GUI=ON
+      - -DCMAKE_C_COMPILER=clang
+      - -DCMAKE_CXX_COMPILER=clang++
+      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
+    sources:
+      # We're using a git source here rather than a zipball because zip/tarballs
+      # don't include submodules, which shadps4 depends on (in the externals directory)
+      - type: git
+        url: https://github.com/shadps4-emu/shadPS4
+        tag: v.0.4.0
```
Comment (jardon) [inline net.shadps4.shadps4.yml:101]: how does this work with the releases xml?

Comment (fpiesche) [reply net.shadps4.shadps4.yml:101]: It does not - the releases.xml is a manually maintained file and used mostly for information. For example, Flathub uses it to generate the "what's new in [version]" segment on its web listings for apps.

I'm working on, but haven't finished, a GitHub Action that can be run when a new GitHub release is created to plug the information from the GH release into the .releases.xml (or the metainfo.xml if required) and update the release tag to a new commit with the updated file. Once I've finished and tested this I'll be happy to PR a workflow to automatically maintain the releases.xml file :)

File: net.shadps4.shadps4.yml (hunk 4 above)
Comment (fpiesche) [reply net.shadps4.shadps4.yml:1]: I would happily do this and there's no hard rules for this; however the FreeDesktop specification "strongly encourages" using only lowercase letters. One reason for this is that some uses of the ID are case sensitive; eg. the `flatpak` command (so eg. `flatpak install org.wesnoth.wesnoth` will not work because the package is named `org.wesnoth.Wesnoth`). With that in mind, would you still prefer the ID to match the normal `shadPS4` stylisation?

Comment (ngoquang2708) [reply net.shadps4.shadps4.yml:1]: As pointed out by the [Flatpak documentation#Requirements & Conventions](https://docs.flatpak.org/en/latest/conventions.html):
>As will be seen below and in future sections, this ID is expected to be used in a number of places. Developers must follow the standard [D-Bus naming conventions for bus names](https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-names) when creating their own IDs. This format is already recommended by the [Desktop File specification](https://specifications.freedesktop.org/desktop-entry-spec/latest/file-naming.html) and also the [Appstream specification](https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#sect-Metadata-GenericComponent).

From [D-Bus naming conventions for bus names](https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-names) :
>Interface names should start with the reversed DNS domain name of the author of the interface (in lower-case), like interface names in Java. It is conventional for the rest of the interface name to consist of words run together, with initial capital letters on all words ("CamelCase").

From [Desktop Entry Specification#file_naming](https://specifications.freedesktop.org/desktop-entry-spec/latest/file-naming.html):
>The name of the desktop entry should follow the "reverse DNS" convention: it should start with a reversed DNS domain name controlled by the author of the application, in lower case. The domain name should be followed by the name of the application, which is conventionally written with words run together and initial capital letters (CamelCase). For example, if the owner of example.org writes "Foo Viewer", they might choose the name org.example.FooViewer, resulting in a file named org.example.FooViewer.desktop. 

There is a section in the  [Flatpak documentation#Requirements & Conventions](https://docs.flatpak.org/en/latest/conventions.html) mentioned that:
>The domain portion of the ID must be in lowercase, and while not required, the application portion is recommended to be in lowercase as well. Therefore, you should use org.example.app.

But in practice, usually, app names use ```CamelCase```:
Screenshot_20241109_100439

In conclusion, using ```net.shadps4.shadPS4``` as ```app-id``` and naming files is the best choice.

Comment (fpiesche) [reply net.shadps4.shadps4.yml]: Done!

**Comment by fpiesche** [comment]
The 6.8 KDE runtime was rebuilt on Friday so should now have libfmt:

- [`org.freedesktop.Platform` v24.08.6 was built on Thursday, including the libfmt change](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/releases/freedesktop-sdk-24.08.6)
- `org.kde.Platform` was rebuilt on Friday, likely as a downstream build from the freedesktop 24.08.6 release:

```
KDE Application Platform - Shared libraries used by KDE applications

          ID: org.kde.Platform
         Ref: runtime/org.kde.Platform/x86_64/6.8
        Arch: x86_64
      Branch: 6.8
     License: GPL-2.0+
      Origin: flathub
  Collection: org.flathub.Stable
Installation: user
   Installed: 1.0 GB

      Commit: a8f6c0c5ae504721f64ef55689e0ffb603b2c65a161edf2d59b7fa6433be8687
      Parent: 247722ce3a44978223909cc2f448ccca48efed9c8e73b56d98e6b12b9022e7ee
     Subject: build of org.kde.Sdk, Fri Nov  8 02:05:06 UTC 2024 (d09f5c1159e9d3c387ed5ebc9591dcb612ae5ac9)
        Date: 2024-11-08 10:37:05 +0000
```

File: net.shadps4.shadPS4.yml [hunk 8]
```diff
@@ -0,0 +1,101 @@
+app-id: net.shadps4.shadPS4
+runtime: org.kde.Platform
+runtime-version: '6.8'
+sdk: org.kde.Sdk
+command: shadps4
+
+# TODO: vvvvv These can go away once shadPS4 PRs 1494 and 1495 are merged vvvvv
```
Comment (jardon) [inline net.shadps4.shadPS4.yml:7]: this needs to be updated.  im assuming the icon can be removed from this PR as well?

**Review by jardon — CHANGES_REQUESTED** [review]

**Comment by ngoquang2708** [comment]
A bit nitpick but manifest file extension should be in full form ```yaml```.

File: net.shadps4.shadPS4.yml [hunk 9]
```diff
@@ -81,21 +76,10 @@ modules:
...
         x-checker-data:
           type: json
           url: https://api.github.com/repos/shadps4-emu/shadps4/releases/latest
           tag-query: .tag_name
           version-query: .tag_name | sub("^v."; "")
-    # TODO: vvvvv These can go away once shadPS4 PRs 1494 and 1495 are merged vvvvv
-      - type: file
-        path: net.shadps4.shadPS4.metainfo.xml
-      - type: file
-        path: net.shadps4.shadPS4.releases.xml
```
Comment (ngoquang2708) [inline net.shadps4.shadPS4.yml]: Since PRs 1494 and 1495 are merged, do we need these files anymore?

**Comment by ngoquang2708** [comment]
Do we need these files here since we already had them in main repo?
```
net.shadps4.shadPS4.metainfo.xml
net.shadps4.shadPS4.releases.xml
net.shadps4.shadPS4_metadata.pot
```

File: net.shadps4.shadPS4.yml (hunk 8 above)
Comment (fpiesche) [reply net.shadps4.shadPS4.yml]: Sorted that - added a source for the icon from the main repo :)

File: net.shadps4.shadPS4.yml [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+      - -DENABLE_QT_GUI=ON
+      - -DCMAKE_C_COMPILER=clang
+      - -DCMAKE_CXX_COMPILER=clang++
+      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
+    sources:
+      # We're using a git source here rather than a zipball because zip/tarballs
+      # don't include submodules, which shadps4 depends on (in the externals directory)
+      - type: git
+        url: https://github.com/shadps4-emu/shadPS4
+        tag: v.0.4.0
```
[the remark this answers is not in the document]
Comment (fpiesche) [reply net.shadps4.shadPS4.yml]: The main/"stable" branch of Flathub applications should generally be tagged releases where available; my best guess is that rpcs3 was added before that guideline was put in place. I'll create a separate PR for a beta branch using `main`.

**Review by ngoquang2708 — APPROVED** [review]
